src/persist-client/src/internal/machine.rs

```diff
@@ -772,356 +772,346 @@ where
-                            let read = client
-                                .open_reader::<String, (), u64, i64>(shard_id)
-                                .await
-                                .expect("invalid shard types");
-                            let listen = read.expect_listen(as_of).await;
-                            state.listens.insert(output.to_owned(), listen);
-                            "ok\n".into()
-                        }
-                        "listen-through" => {
-                            let input = get_arg(&tc.args, "input").expect("missing input");
-                            let frontier = get_u64(&tc.args, "frontier").expect("missing frontier");
-                            let listen = state.listens.get_mut(input).expect("unknown listener");
-                            let mut s = String::new();
-                            'outer: loop {
-                                for event in listen.next().await {
-                                    match event {
-                                        ListenEvent::Updates(x) => {
-                                            for ((k, _v), t, d) in x.iter() {
-                                                write!(s, "{} {} {}\n", k.as_ref().unwrap(), t, d);
-                                            }
-                                        }
-                                        ListenEvent::Progress(x) => {
-                                            if !x.less_than(&frontier) {
-                                                break 'outer;
-                                            }
-                                        }
-                                    }
-                                }
-                            }
-                            if s.is_empty() {
-                                s.push_str("<empty>\n");
-                            }
-                            s
-                        }
-                        "compare-and-append" => {
-                            let input = get_arg(&tc.args, "input").expect("missing input");
-                            let batch = state.batches.get(input).expect("unknown batch");
-                            let mut write = write.lock().await;
-                            let writer_id = write.writer_id.clone();
-                            let (_, _) = write
-                                .machine
-                                .compare_and_append(batch, &writer_id, now())
-                                .await
-                                .expect("indeterminate")
-                                .expect("invalid usage")
-                                .expect("upper mismatch");
-                            format!("ok\n")
+    /// Scans consensus and returns all states with their SeqNos
+    /// and which batches they reference
+    pub async fn consensus_scan(
+        dd: &mut MachineState,
```

Comment:
whoops missed some `dd`s

Reply:
honestly just `datadriven` seems fine to me. it's verbose but very clear

Reply:
done!